Improve error for AddEntityFrameworkStores

HaoK opened the pull request:
Fixes https://github.com/aspnet/Identity/issues/1001

@divega

divega commented on src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/IdentityEntityFrameworkBuilderExtensions.cs:
```diff
@@ -38,26 +39,28 @@ public static IdentityBuilder AddEntityFrameworkStores<TContext>(this IdentityBu
+            var identityUserType = typeof(IdentityUser<>).MakeGenericType(keyType);
+            if (!identityUserType.GetTypeInfo().IsAssignableFrom(userType.GetTypeInfo()))
+            {
+                throw new InvalidOperationException(Resources.NotIdentityUser);
```
AFAIR there were a couple of ways to hit the current `TargetInvocationException`. Need to search around to make sure we are covered by this check.

HaoK commented:
We could just catch all exceptions and rethrow with our message and the original exception as an inner exception to be safe.

divega commented:
@HaoK I looked around, and I am almost sure the two cases I had seen in which we would throw the exception are:
- Inherit from the IdentityUser that has more generic arguments
- Change TKey to a different type from the default but forget to pass the new type as a generic argument in the call to AddEntityFrameworkStores. 

For the latter I noticed you filed https://github.com/aspnet/Identity/issues/876.

divega commented on src/Microsoft.AspNetCore.Identity.EntityFrameworkCore/Resources.resx:
```diff
@@ -117,6 +117,14 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
+  <data name="NotIdentityRole" xml:space="preserve">
+    <value>AddEntityFrameworkStores can only be called with a role that derives from IdentityRole&lt;TKey&gt;. If you are specifying more generic arguments, use IdentityBuilder.AddRoleStore&lt;TStore&gt;() instead.</value>
```
Couple of observations:
- I don't think the mention of `IdentityBuilder` helps here. I would just remove it. 
- Customers also need to derive their own store classes (by overriding a method in the abstract base class), right? If that is the case, can you also mention it in the messages?

divega commented:
BTW, a couple of simpe searches shows that this exception is [one of the most common reasons customers file bugs on the Identity repo](https://github.com/aspnet/Identity/search?q=GenericArguments+&type=Issues&utf8=%E2%9C%93) and [ask question about Identity over in StackOverflow](http://stackoverflow.com/search?q=Identity+GenericArguments). 

If we get it right this could help a lot!

HaoK commented:
I'll try to fix #876 in this PR as well

HaoK commented:
Updated with the error message tweaks, also added support for infering the TKey from any DbContext that dervies from IdentityDbContext<TUser,TRole,TKey>.  This makes the second overload unnecessary but we can keep it for breaking changes, and also supporting custom DbContext's that don't derive from our DbContext (which isn't a requirement today)

HaoK commented:
13ae7b21f3459e97cd5853b6e6ccb229dac1bafe
